File: ios/MitiMaiti/Services/APIService.swift
import Foundation

actor APIService {
    static let shared = APIService()
    private var accessToken: String?
    private var refreshToken: String?
    private let http = HTTPClient.shared
    private let store = TokenStore.shared

    func bootstrap() async {
        let (a, r) = await store.load()
        self.accessToken = a
        self.refreshToken = r
    }

    func setTokens(access: String, refresh: String) async {
        self.accessToken = access
        self.refreshToken = refresh
        await store.save(access: access, refresh: refresh)
    }

    func clearTokens() async {
        self.accessToken = nil
        self.refreshToken = nil
        await store.clear()
    }

    func currentAccessToken() -> String? { accessToken }

    // MARK: - Auth

    func sendOTP(phone: String) async throws -> Bool {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 1_000_000_000)
            return true
        }
        struct Body: Encodable { let phone: String }
        let _: EmptyData = try await http.request(.post, "/auth/login", body: Body(phone: phone))
        return true
    }

    func verifyOTP(phone: String, code: String) async throws -> (accessToken: String, refreshToken: String, isNew: Bool) {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 1_500_000_000)
            guard code == "123456" else { throw APIError.invalidOTP }
            let access = "mock-access-token-\(UUID().uuidString)"
            let refresh = "mock-refresh-token-\(UUID().uuidString)"
            await setTokens(access: access, refresh: refresh)
            return (access, refresh, true)
        }
        struct Body: Encodable { let phone: String; let token: String }
        struct Resp: Decodable {
            struct UserStub: Decodable { let isNew: Bool }
            struct Session: Decodable { let accessToken: String; let refreshToken: String }
            let user: UserStub
            let session: Session
        }
        do {
            let resp: Resp = try await http.request(.post, "/auth/verify", body: Body(phone: phone, token: code))
            await setTokens(access: resp.session.accessToken, refresh: resp.session.refreshToken)
            return (resp.session.accessToken, resp.session.refreshToken, resp.user.isNew)
        } catch APIError.unauthorized {
            throw APIError.invalidOTP
        }
    }

    func refresh() async throws {
        guard let refreshToken else { throw APIError.unauthorized }
        struct Body: Encodable { let refreshToken: String }
        struct Resp: Decodable { let accessToken: String; let refreshToken: String }
        let resp: Resp = try await http.request(.post, "/auth/refresh", body: Body(refreshToken: refreshToken))
        await setTokens(access: resp.accessToken, refresh: resp.refreshToken)
    }

    // MARK: - Photo upload

    func uploadPhoto(imageData: Data, mimeType: String = "image/jpeg") async throws -> UserPhoto {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 800_000_000)
            return UserPhoto(
                url: "https://i.pravatar.cc/600?u=\(UUID().uuidString)",
                isPrimary: false,
                sortOrder: 0
            )
        }
        let boundary = UUID().uuidString
        var body = Data()
        let header = "--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"photo.jpg\"\r\nContent-Type: \(mimeType)\r\n\r\n"
        body.append(header.data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        let url = URL(string: AppConfig.baseURL + "/me/media")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        if let token = accessToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw APIError.networkError }
        guard (200..<300).contains(http.statusCode) else {
            if http.statusCode == 401 {
                try await refresh()
                return try await uploadPhoto(imageData: imageData, mimeType: mimeType)
            }
            throw APIError.serverError("Upload failed: HTTP \(http.statusCode)")
        }

        struct Resp: Decodable {
            struct Media: Decodable {
                let id: String
                let urlThumb: String?
                let urlMedium: String?
                let urlOriginal: String
                let isPrimary: Bool?
                let sortOrder: Int?
            }
            let media: Media
        }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let envelope = try decoder.decode(APIEnvelope<Resp>.self, from: data)
        guard let resp = envelope.data else { throw APIError.serverError("Empty response") }
        return UserPhoto(
            id: resp.media.id,
            url: resp.media.urlOriginal,
            urlThumb: resp.media.urlThumb,
            urlMedium: resp.media.urlMedium,
            isPrimary: resp.media.isPrimary ?? false,
            sortOrder: resp.media.sortOrder ?? 0
        )
    }

    // MARK: - Profile

    func fetchProfile() async throws -> User {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 800_000_000)
            return MockData.currentUser
        }
        let resp: ProfileResponse = try await authedRequest(.get, "/me")
        return resp.toUser()
    }

    func updateProfile(_ updates: [String: Any]) async throws -> User {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 600_000_000)
            return MockData.currentUser
        }
        struct Resp: Decodable { let user: User }
        let body = try JSONSerialization.data(withJSONObject: updates)
        let resp: Resp = try await authedRequest(.patch, "/me", rawBody: body)
        return resp.user
    }

    // MARK: - Feed

    func fetchFeed(cursor: String? = nil) async throws -> [FeedCard] {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 1_000_000_000)
            return MockData.feedCards
        }
        struct Resp: Decodable { let cards: [FeedCard] }
        let path = cursor.map { "/feed?cursor=\($0)" } ?? "/feed"
        let resp: Resp = try await authedRequest(.get, path)
        return resp.cards
    }

    // MARK: - Actions

    func performAction(targetId: String, type: ActionType) async throws -> (isMatch: Bool, matchId: String?) {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 500_000_000)
            let isMatch = Double.random(in: 0...1) > 0.7
            return (isMatch, isMatch ? UUID().uuidString : nil)
        }
        struct Body: Encodable { let targetId: String; let type: String }
        struct Resp: Decodable { let isMatch: Bool; let matchId: String? }
        let resp: Resp = try await authedRequest(.post, "/action", body: Body(targetId: targetId, type: String(describing: type)))
        return (resp.isMatch, resp.matchId)
    }

    func registerFcmToken(_ token: String, platform: String = "ios") async throws {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 200_000_000)
            return
        }
        struct Body: Encodable { let token: String; let platform: String }
        let _: EmptyData = try await authedRequest(.post, "/me/fcm-token", body: Body(token: token, platform: platform))
    }

    func answerPrompt(_ answer: String) async throws {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 400_000_000)
            return
        }
        struct Body: Encodable { let answer: String }
        let _: EmptyData = try await authedRequest(.post, "/action/prompt", body: Body(answer: answer))
    }

    func joinFamily(code: String, roleTag: String) async throws {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 500_000_000)
            return
        }
        struct Body: Encodable { let code: String; let roleTag: String }
        let _: EmptyData = try await authedRequest(.post, "/family/join", body: Body(code: code, roleTag: roleTag))
    }

    func rewind() async throws -> String {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 500_000_000)
            return UUID().uuidString
        }
        struct Resp: Decodable { let restoredId: String }
        let resp: Resp = try await authedRequest(.post, "/action/rewind")
        return resp.restoredId
    }

    // MARK: - Inbox

    func fetchInbox() async throws -> (likes: [LikedYouCard], matches: [Match]) {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 800_000_000)
            return (MockData.mockLikes, MockData.mockMatches)
        }
        struct Resp: Decodable { let likes: [LikedYouCard]; let matches: [Match] }
        let resp: Resp = try await authedRequest(.get, "/inbox")
        return (resp.likes, resp.matches)
    }

    // MARK: - Chat

    func fetchMessages(matchId: String, before: String? = nil) async throws -> [Message] {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 600_000_000)
            let otherUserId = MockData.mockMatches.first(where: { $0.id == matchId })?.otherUser.id ?? "other"
            return MockData.mockMessages(matchId: matchId, otherUserId: otherUserId)
        }
        struct Resp: Decodable { let messages: [Message] }
        var path = "/chat/\(matchId)"
        if let before { path += "?before=\(before)" }
        let resp: Resp = try await authedRequest(.get, path)
        return resp.messages
    }

    func sendChatMedia(matchId: String, imageData: Data, mimeType: String = "image/jpeg") async throws -> Message {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 600_000_000)
            return Message(matchId: matchId, senderId: "current-user-id",
                           content: "https://i.pravatar.cc/600?u=\(UUID().uuidString)",
                           msgType: .photo, status: .sent)
        }
        let boundary = UUID().uuidString
        var body = Data()
        let header = "--\(boundary)\r\nContent-Disposition: form-data; name=\"media\"; filename=\"chat.jpg\"\r\nContent-Type: \(mimeType)\r\n\r\n"
        body.append(header.data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        let url = URL(string: AppConfig.baseURL + "/chat/\(matchId)/media")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        if let token = accessToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw APIError.networkError }
        guard (200..<300).contains(http.statusCode) else {
            if http.statusCode == 401 {
                try await refresh()
                return try await sendChatMedia(matchId: matchId, imageData: imageData, mimeType: mimeType)
            }
            throw APIError.serverError("Chat media upload failed: HTTP \(http.statusCode)")
        }

        struct Resp: Decodable { let message: Message }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
        let envelope = try decoder.decode(APIEnvelope<Resp>.self, from: data)
        guard let resp = envelope.data else { throw APIError.serverError("Empty response") }
        return resp.message
    }

    func sendMessage(matchId: String, content: String, type: MessageType = .text) async throws -> Message {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 300_000_000)
            return Message(matchId: matchId, senderId: "current-user-id", content: content, msgType: type, status: .sent)
        }
        struct Body: Encodable { let content: String; let type: String }
        struct Resp: Decodable { let message: Message }
        let resp: Resp = try await authedRequest(.post, "/chat/\(matchId)/messages", body: Body(content: content, type: String(describing: type)))
        return resp.message
    }

    // MARK: - Family

    func fetchFamily() async throws -> (members: [FamilyMember], suggestions: [FamilySuggestion]) {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 600_000_000)
            return (MockData.mockFamilyMembers, MockData.mockFamilySuggestions)
        }
        struct MembersResp: Decodable { let members: [FamilyMember] }
        struct SuggestionsResp: Decodable { let suggestions: [FamilySuggestion] }
        async let members: MembersResp = authedRequest(.get, "/family")
        async let suggestions: SuggestionsResp = authedRequest(.get, "/family/suggestions")
        let (m, s) = try await (members, suggestions)
        return (m.members, s.suggestions)
    }

    func generateInvite() async throws -> FamilyInvite {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 500_000_000)
            return FamilyInvite()
        }
        struct Resp: Decodable { let invite: FamilyInvite }
        let resp: Resp = try await authedRequest(.post, "/family/invite")
        return resp.invite
    }

    // MARK: - Safety

    func reportUser(userId: String, reason: String, details: String?) async throws {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 500_000_000)
            return
        }
        struct Body: Encodable { let targetUserId: String; let reason: String; let details: String? }
        let _: EmptyData = try await authedRequest(.post, "/safety/report", body: Body(targetUserId: userId, reason: reason, details: details))
    }

    func blockUser(userId: String) async throws {
        if AppConfig.useMockData {
            try await Task.sleep(nanoseconds: 500_000_000)
            return
        }
        struct Body: Encodable { let targetUserId: String }
        let _: EmptyData = try await authedRequest(.post, "/safety/block", body: Body(targetUserId: userId))
    }

    // MARK: - Authed request helper with 401 → refresh → retry

    private func authedRequest<T: Decodable>(
        _ method: HTTPMethod,
        _ path: String,
        body: Encodable? = nil,
        rawBody: Data? = nil
    ) async throws -> T {
        do {
            return try await http.request(method, path, body: body, rawBody: rawBody, accessToken: accessToken)
        } catch APIError.unauthorized {
            try await refresh()
            return try await http.request(method, path, body: body, rawBody: rawBody, accessToken: accessToken)
        }
    }
}

// MARK: - API Errors

enum APIError: LocalizedError {
    case invalidOTP
    case networkError
    case unauthorized
    case rateLimited
    case serverError(String)

    var errorDescription: String? {
        switch self {
        case .invalidOTP: return "Invalid verification code. Please try again."
        case .networkError: return "Network error. Please check your connection."
        case .unauthorized: return "Session expired. Please log in again."
        case .rateLimited: return "Too many requests. Please wait a moment."
        case .serverError(let msg): return msg
        }
    }
}

File: android/app/src/main/java/com/mitimaiti/app/services/APIService.kt
package com.mitimaiti.app.services

import com.mitimaiti.app.models.*
import kotlinx.coroutines.delay
import java.time.LocalDate

/**
 * Unified API service that switches between mock data and real backend.
 * Toggle via ApiConfig.useMockData
 *
 * Real backend endpoints (backend/src/routes/):
 *   POST /v1/auth/login        → sendOTP
 *   POST /v1/auth/verify       → verifyOTP
 *   POST /v1/auth/refresh      → refreshToken
 *   POST /v1/auth/delete       → logout / deleteAccount
 *   GET  /v1/me                → fetchProfile
 *   PATCH /v1/me               → updateProfile
 *   GET  /v1/feed              → fetchFeed
 *   POST /v1/action            → performAction (like/pass)
 *   POST /v1/action/rewind     → rewind
 *   GET  /v1/inbox             → fetchInbox
 *   GET  /v1/chat/:matchId     → fetchMessages
 *   POST /v1/chat/:matchId/messages → sendMessage (REST fallback)
 *   POST /v1/family/invite     → generateInvite
 *   GET  /v1/family            → fetchFamily
 *   POST /v1/safety/report     → reportUser
 *   POST /v1/safety/block      → blockUser
 *
 * Real-time chat uses SocketManager (Socket.IO on port 4001)
 */
object APIService {
    private var tokenManager: TokenManager? = null
    private val api: MitiMaitiApi by lazy { HttpClient.retrofit.create(MitiMaitiApi::class.java) }

    private var accessToken: String? = null
    private var refreshToken: String? = null

    fun init(tokenManager: TokenManager) {
        this.tokenManager = tokenManager
        HttpClient.init(tokenManager)
    }

    fun setTokens(access: String, refresh: String) { accessToken = access; refreshToken = refresh }
    fun clearTokens() { accessToken = null; refreshToken = null }

    // ──────────────────── AUTH ────────────────────

    suspend fun sendOTP(phone: String): Result<Boolean> {
        if (ApiConfig.useMockData) {
            delay(1000)
            return if (phone.length >= 10) Result.success(true) else Result.failure(APIError.NetworkError)
        }
        return try {
            val response = api.sendOTP(mapOf("phone" to phone))
            if (response.isSuccessful) Result.success(true)
            else Result.failure(APIError.NetworkError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    suspend fun verifyOTP(phone: String, code: String): Result<Pair<User, Boolean>> {
        if (ApiConfig.useMockData) {
            delay(1500)
            return if (code == "123456") {
                setTokens("mock-access-token", "mock-refresh-token")
                Result.success(Pair(MockData.currentUser, true))
            } else Result.failure(APIError.InvalidOTP)
        }
        return try {
            val response = api.verifyOTP(mapOf("phone" to phone, "token" to code))
            if (response.isSuccessful) {
                val body = response.body() ?: return Result.failure(APIError.ServerError)
                val token = body["access_token"] as? String ?: ""
                val refresh = body["refresh_token"] as? String ?: ""
                val userId = body["user_id"] as? String ?: ""
                val isNew = body["is_new_user"] as? Boolean ?: false
                setTokens(token, refresh)
                tokenManager?.saveTokens(token, refresh, userId)
                // Connect WebSocket after auth
                SocketManager.shared.connect(token)
                Result.success(Pair(parseUser(body["user"] as? Map<*, *>), isNew))
            } else {
                val code = response.code()
                if (code == 401) Result.failure(APIError.InvalidOTP)
                else Result.failure(APIError.ServerError)
            }
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── PROFILE ────────────────────

    suspend fun fetchProfile(): Result<User> {
        if (ApiConfig.useMockData) { delay(500); return Result.success(MockData.currentUser) }
        return try {
            val response = api.getProfile()
            if (response.isSuccessful) {
                Result.success(parseUser(response.body()))
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    suspend fun updateProfile(updates: Map<String, Any>): Result<User> {
        if (ApiConfig.useMockData) { delay(800); return Result.success(MockData.currentUser) }
        return try {
            val response = api.updateProfile(updates)
            if (response.isSuccessful) Result.success(parseUser(response.body()))
            else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── FEED ────────────────────

    suspend fun fetchFeed(cursor: String? = null): Result<List<FeedCard>> {
        if (ApiConfig.useMockData) { delay(1000); return Result.success(MockData.makeFeedCards()) }
        return try {
            val response = api.getFeed(cursor)
            if (response.isSuccessful) {
                val body = response.body() ?: return Result.success(emptyList())
                val cards = parseFeedCards(body["cards"] as? List<*>)
                Result.success(cards)
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── ACTIONS ────────────────────

    suspend fun performAction(targetId: String, type: String): Result<Match?> {
        if (ApiConfig.useMockData) {
            delay(500)
            if (type == "like") {
                val isMatch = (1..10).random() <= 3
                if (isMatch) {
                    val user = MockData.mockUsers.firstOrNull { it.id == targetId } ?: MockData.mockUsers.first()
                    return Result.success(Match(otherUser = user, status = MatchStatus.PENDING_FIRST_MESSAGE, matchedAt = System.currentTimeMillis(), expiresAt = System.currentTimeMillis() + 24 * 60 * 60 * 1000L))
                }
            }
            return Result.success(null)
        }
        return try {
            val response = api.performAction(mapOf("targetUserId" to targetId, "type" to type))
            if (response.isSuccessful) {
                val body = response.body()
                val matchData = body?.get("match") as? Map<*, *>
                if (matchData != null) Result.success(parseMatch(matchData))
                else Result.success(null)
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    suspend fun rewind(): Result<Boolean> {
        if (ApiConfig.useMockData) { delay(300); return Result.success(true) }
        return try {
            val response = api.rewind()
            Result.success(response.isSuccessful)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── INBOX ────────────────────

    suspend fun fetchInbox(): Result<Pair<List<LikedYouCard>, List<Match>>> {
        if (ApiConfig.useMockData) { delay(800); return Result.success(Pair(MockData.makeLikes(), MockData.makeMatches())) }
        return try {
            val response = api.getInbox()
            if (response.isSuccessful) {
                val body = response.body() ?: return Result.success(Pair(emptyList(), emptyList()))
                val likes = parseLikes(body["liked_you"] as? List<*>)
                val matches = parseMatches(body["matches"] as? List<*>)
                Result.success(Pair(likes, matches))
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── CHAT ────────────────────

    suspend fun fetchMessages(matchId: String): Result<List<Message>> {
        if (ApiConfig.useMockData) { delay(500); return Result.success(MockData.makeMessages(matchId)) }
        return try {
            val response = api.getMessages(matchId)
            if (response.isSuccessful) {
                val body = response.body() ?: return Result.success(emptyList())
                val messages = parseMessages(body["messages"] as? List<*>)
                Result.success(messages)
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    suspend fun sendMessage(matchId: String, content: String, type: MessageType = MessageType.TEXT): Result<Message> {
        if (ApiConfig.useMockData) {
            delay(300)
            return Result.success(Message(matchId = matchId, senderId = "current-user-id", content = content, msgType = type, status = MessageStatus.SENT))
        }
        // Prefer WebSocket for real-time, REST as fallback
        if (SocketManager.shared.isConnected.value) {
            SocketManager.shared.sendMessage(matchId, content, type.value)
            return Result.success(Message(matchId = matchId, senderId = tokenManager?.getUserId().toString(), content = content, msgType = type, status = MessageStatus.SENDING))
        }
        return try {
            val response = api.sendMessage(matchId, mapOf("content" to content, "msgType" to type.value))
            if (response.isSuccessful) {
                val body = response.body()
                Result.success(parseMessage(body?.get("message") as? Map<*, *>))
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── FAMILY ────────────────────

    suspend fun fetchFamily(): Result<Pair<List<FamilyMember>, List<FamilySuggestion>>> {
        if (ApiConfig.useMockData) { delay(600); return Result.success(Pair(MockData.makeFamilyMembers(), MockData.makeFamilySuggestions())) }
        return try {
            val response = api.getFamily()
            if (response.isSuccessful) {
                val body = response.body() ?: return Result.success(Pair(emptyList(), emptyList()))
                val members = parseFamilyMembers(body["members"] as? List<*>)
                val suggestions = parseFamilySuggestions(body["suggestions"] as? List<*>)
                Result.success(Pair(members, suggestions))
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    suspend fun generateInvite(): Result<FamilyInvite> {
        if (ApiConfig.useMockData) {
            delay(400)
            val code = "MM-${(100000..999999).random()}"
            return Result.success(FamilyInvite(code = code, deepLink = "https://mitimaiti.com/join/$code", currentMembers = MockData.makeFamilyMembers().size))
        }
        return try {
            val response = api.generateFamilyInvite()
            if (response.isSuccessful) {
                val body = response.body() ?: return Result.failure(APIError.ServerError)
                Result.success(FamilyInvite(
                    code = body["code"] as? String ?: "",
                    deepLink = body["deep_link"] as? String ?: "",
                    currentMembers = (body["current_members"] as? Double)?.toInt() ?: 0
                ))
            } else Result.failure(APIError.ServerError)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── SAFETY ────────────────────

    suspend fun reportUser(userId: String, reason: String): Result<Boolean> {
        if (ApiConfig.useMockData) { delay(500); return Result.success(true) }
        return try {
            val response = api.reportUser(mapOf("reported_user_id" to userId, "reason" to reason))
            Result.success(response.isSuccessful)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    suspend fun blockUser(userId: String): Result<Boolean> {
        if (ApiConfig.useMockData) { delay(500); return Result.success(true) }
        return try {
            val response = api.blockUser(mapOf("blocked_user_id" to userId))
            Result.success(response.isSuccessful)
        } catch (e: Exception) { Result.failure(APIError.NetworkError) }
    }

    // ──────────────────── PARSERS ────────────────────

    private fun parseUser(data: Map<*, *>?): User {
        if (data == null) return MockData.currentUser
        return User(
            id = data["id"] as? String ?: "",
            phone = data["phone"] as? String ?: "",
            displayName = data["display_name"] as? String ?: "",
            dateOfBirth = (data["date_of_birth"] as? String)?.let { try { LocalDate.parse(it) } catch (e: Exception) { null } },
            gender = (data["gender"] as? String)?.let { g -> Gender.entries.firstOrNull { it.name.equals(g, true) } },
            bio = data["bio"] as? String ?: "",
            heightCm = (data["height_cm"] as? Double)?.toInt(),
            city = data["city"] as? String ?: "",
            state = data["state"] as? String ?: "",
            country = data["country"] as? String ?: "",
            intent = (data["intent"] as? String)?.let { i -> Intent.entries.firstOrNull { it.name.equals(i, true) } },
            isVerified = data["is_verified"] as? Boolean ?: false,
            photos = (data["photos"] as? List<*>)?.map { p ->
                val photo = p as? Map<*, *> ?: return@map UserPhoto(url = "")
                UserPhoto(
                    id = photo["id"] as? String ?: "",
                    url = photo["url_original"] as? String ?: photo["url"] as? String ?: "",
                    urlThumb = photo["url_thumb"] as? String,
                    urlMedium = photo["url_medium"] as? String,
                    isPrimary = photo["is_primary"] as? Boolean ?: false,
                    sortOrder = (photo["sort_order"] as? Double)?.toInt() ?: 0,
                    isVerified = photo["is_verified"] as? Boolean ?: false
                )
            } ?: emptyList(),
            occupation = data["occupation"] as? String,
            education = data["education"] as? String,
            company = data["company"] as? String,
            religion = data["religion"] as? String,
            sindhiFluency = (data["sindhi_fluency"] as? String)?.let { f -> SindhiFluency.entries.firstOrNull { it.name.equals(f, true) } },
            familyValues = (data["family_values"] as? String)?.let { f -> FamilyValues.entries.firstOrNull { it.name.equals(f, true) } },
            foodPreference = (data["food_preference"] as? String)?.let { f -> FoodPreference.entries.firstOrNull { it.name.equals(f, true) } },
            interests = (data["interests"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList(),
            isOnline = data["is_online"] as? Boolean ?: false,
            profileCompleteness = (data["profile_completeness"] as? Double)?.toInt() ?: 0
        )
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseFeedCards(data: List<*>?): List<FeedCard> {
        return data?.mapNotNull { item ->
            val card = item as? Map<*, *> ?: return@mapNotNull null
            val score = (card["cultural_score"] as? Double)?.toInt() ?: 0
            FeedCard(
                id = card["id"] as? String ?: "",
                user = parseUser(card as? Map<*, *>),
                culturalScore = CulturalScore(
                    overallScore = score,
                    badge = when { score >= 85 -> CulturalBadge.GOLD; score >= 65 -> CulturalBadge.GREEN; score >= 40 -> CulturalBadge.ORANGE; else -> CulturalBadge.NONE }
                ),
                commonInterests = (card["common_interests"] as? Double)?.toInt() ?: 0,
                distanceKm = card["distance_km"] as? Double
            )
        } ?: emptyList()
    }

    private fun parseLikes(data: List<*>?): List<LikedYouCard> {
        return data?.mapNotNull { item ->
            val like = item as? Map<*, *> ?: return@mapNotNull null
            val score = (like["cultural_score"] as? Double)?.toInt() ?: 0
            LikedYouCard(
                id = like["id"] as? String ?: "",
                user = parseUser(like["user"] as? Map<*, *>),
                culturalScore = score,
                culturalBadge = when { score >= 85 -> CulturalBadge.GOLD; score >= 65 -> CulturalBadge.GREEN; else -> CulturalBadge.ORANGE }
            )
        } ?: emptyList()
    }

    private fun parseMatches(data: List<*>?): List<Match> {
        return data?.mapNotNull { item -> parseMatch(item as? Map<*, *>) } ?: emptyList()
    }

    private fun parseMatch(data: Map<*, *>?): Match {
        if (data == null) return Match(otherUser = MockData.mockUsers.first())
        return Match(
            id = data["id"] as? String ?: "",
            otherUser = parseUser(data["other_user"] as? Map<*, *>),
            status = (data["status"] as? String)?.let { s -> MatchStatus.entries.firstOrNull { it.value == s } } ?: MatchStatus.PENDING_FIRST_MESSAGE,
            matchedAt = (data["matched_at"] as? Double)?.toLong() ?: System.currentTimeMillis(),
            expiresAt = (data["expires_at"] as? Double)?.toLong(),
            lastMessage = data["last_message"] as? String,
            unreadCount = (data["unread_count"] as? Double)?.toInt() ?: 0,
            firstMsgBy = data["first_msg_by"] as? String,
            firstMsgLocked = data["first_msg_locked"] as? Boolean ?: false,
            firstMsgAt = (data["first_msg_at"] as? Double)?.toLong()
        )
    }

    private fun parseMessages(data: List<*>?): List<Message> {
        return data?.mapNotNull { item -> parseMessage(item as? Map<*, *>) } ?: emptyList()
    }

    private fun parseMessage(data: Map<*, *>?): Message {
        if (data == null) return Message()
        return Message(
            id = data["id"] as? String ?: "",
            matchId = data["match_id"] as? String ?: "",
            senderId = data["sender_id"] as? String ?: "",
            content = data["content"] as? String ?: "",
            mediaUrl = data["media_url"] as? String,
            msgType = (data["msg_type"] as? String)?.let { t -> MessageType.entries.firstOrNull { it.value == t } } ?: MessageType.TEXT,
            status = if (data["is_read"] as? Boolean == true) MessageStatus.READ else MessageStatus.DELIVERED,
            createdAt = (data["created_at"] as? Double)?.toLong() ?: System.currentTimeMillis()
        )
    }

    private fun parseFamilyMembers(data: List<*>?): List<FamilyMember> {
        return data?.mapNotNull { item ->
            val m = item as? Map<*, *> ?: return@mapNotNull null
            val perms = m["permissions"] as? Map<*, *>
            FamilyMember(
                id = m["id"] as? String ?: "",
                name = m["name"] as? String ?: "",
                phone = m["phone"] as? String ?: "",
                relationship = m["role_tag"] as? String ?: "",
                status = (m["status"] as? String)?.let { s -> FamilyMemberStatus.entries.firstOrNull { it.value == s } } ?: FamilyMemberStatus.ACTIVE,
                permissions = FamilyPermissions(
                    canViewProfile = perms?.get("canViewProfile") as? Boolean ?: true,
                    canViewPhotos = perms?.get("canViewPhotos") as? Boolean ?: true,
                    canViewBasics = perms?.get("canViewBasics") as? Boolean ?: true,
                    canViewSindhi = perms?.get("canViewSindhi") as? Boolean ?: true,
                    canViewMatches = perms?.get("canViewMatches") as? Boolean ?: false,
                    canSuggest = perms?.get("canSuggest") as? Boolean ?: true,
                    canViewCulturalScore = perms?.get("canViewCulturalScore") as? Boolean ?: true,
                    canViewKundli = perms?.get("canViewKundli") as? Boolean ?: false
                )
            )
        } ?: emptyList()
    }

    private fun parseFamilySuggestions(data: List<*>?): List<FamilySuggestion> {
        return data?.mapNotNull { item ->
            val s = item as? Map<*, *> ?: return@mapNotNull null
            FamilySuggestion(
                id = s["id"] as? String ?: "",
                suggestedBy = parseFamilyMembers(listOf(s["suggested_by"])).firstOrNull() ?: return@mapNotNull null,
                suggestedUser = parseUser(s["suggested_user"] as? Map<*, *>),
                note = s["note"] as? String
            )
        } ?: emptyList()
    }
}

sealed class APIError : Exception() {
    object InvalidOTP : APIError()
    object NetworkError : APIError()
    object Unauthorized : APIError()
    object RateLimited : APIError()
    object ServerError : APIError()
}
